
## Week 07- Spark Application


In [0]:
import zipfile
import os

# Copy the .zip file from DBFS to the local filesystem (temporary directory)
dbutils.fs.cp("dbfs:/FileStore/tables/PGYR2023_P01302025_01212025-1.zip", "file:/tmp/PGYR2023_P01302025_01212025-1.zip")

# Unzip the copied file
with zipfile.ZipFile("/tmp/PGYR2023_P01302025_01212025-1.zip", "r") as zip_ref:
    zip_ref.extractall("/tmp/")

# List the contents of the /tmp/ directory
extracted_files = os.listdir("/tmp/")
print(extracted_files)


['hsperfdata_root', 'systemd-private-04abd221ea9c45569792aebf11b18e72-systemd-resolved.service-xgoqri', 'custom-spark.conf', 'OP_DTL_OWNRSHP_PGYR2023_P01302025_01212025.csv', '.Test-unix', 'chauffeur-daemon.pid', 'systemd-private-04abd221ea9c45569792aebf11b18e72-systemd-logind.service-Ou1hOg', 'chauffeur-env.sh', 'systemd-private-04abd221ea9c45569792aebf11b18e72-ntp.service-kRMxEg', 'driver-daemon-params', 'Rtmph2Y4fg', '.ICE-unix', 'driver-env.sh', '.PGYR2023_P01302025_01212025-1.zip.crc', '.font-unix', 'systemd-private-04abd221ea9c45569792aebf11b18e72-apache2.service-moi4ri', 'tmp.gYEsTitdlA', '.XIM-unix', 'OP_PGYR2023_README_P01302025.txt', 'Rserv', 'OP_DTL_RSRCH_PGYR2023_P01302025_01212025.csv', 'driver-daemon.pid', 'OP_REMOVED_DELETED_PGYR2023_P01302025_01212025.csv', 'chauffeur-daemon-params', 'OP_DTL_GNRL_PGYR2023_P01302025_01212025.csv', '.X11-unix', 'PGYR2023_P01302025_01212025-1.zip']


In [0]:
# Define the local file path (temporary location) and DBFS destination path
local_file_path = "file:/tmp/OP_DTL_GNRL_PGYR2023_P01302025_01212025.csv"  # Path in /tmp/ or local system
dbfs_dest_path = "dbfs:/FileStore/tables/OP_DTL_GNRL_PGYR2023_P01302025_01212025.csv"  # DBFS destination path

# Copy the file from the local file system (temporary location) to DBFS
dbutils.fs.cp(local_file_path, dbfs_dest_path)

# Verify that the file is copied to DBFS
print(f"File is copied to DBFS at: {dbfs_dest_path}")


File is copied to DBFS at: dbfs:/FileStore/tables/OP_DTL_GNRL_PGYR2023_P01302025_01212025.csv


In [0]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, desc
from pyspark.sql.types import IntegerType
from pyspark.sql.functions import sum
from pyspark.sql.types import LongType

**Load the data**

CMS 2023 Open Payments file (https://download.cms.gov/openpayments/PGYR2023_P01302025_01212025.zipLinks)

In [0]:
# File location and type
file_location = "dbfs:/FileStore/tables/OP_DTL_GNRL_PGYR2023_P01302025_01212025.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df.limit(5))

Change_Type Covered_Recipient_Type Teaching_Hospital_CCN Teaching_Hospital_ID Teaching_Hospital_Name Covered_Recipient_Profile_ID Covered_Recipient_NPI Covered_Recipient_First_Name Covered_Recipient_Middle_Name Covered_Recipient_Last_Name Covered_Recipient_Name_Suffix Recipient_Primary_Business_Street_Address_Line1 Recipient_Primary_Business_Street_Address_Line2 Recipient_City Recipient_State Recipient_Zip_Code Recipient_Country Recipient_Province Recipient_Postal_Code Covered_Recipient_Primary_Type_1 Covered_Recipient_Primary_Type_2 Covered_Recipient_Primary_Type_3 Covered_Recipient_Primary_Type_4 Covered_Recipient_Primary_Type_5 Covered_Recipient_Primary_Type_6 Covered_Recipient_Specialty_1 Covered_Recipient_Specialty_2 Covered_Recipient_Specialty_3 Covered_Recipient_Specialty_4 Covered_Recipient_Specialty_5 Covered_Recipient_Specialty_6 Covered_Recipient_License_State_code1 Covered_Recipient_License_State_code2 Covered_Recipient_License_State_code3 Covered_Recipient_License_State_code4 Covered_Recipient_License_State_code5 Submitting_Applicable_Manufacturer_or_Applicable_GPO_Name Applicable_Manufacturer_or_Applicable_GPO_Making_Payment_ID Applicable_Manufacturer_or_Applicable_GPO_Making_Payment_Name Applicable_Manufacturer_or_Applicable_GPO_Making_Payment_State Applicable_Manufacturer_or_Applicable_GPO_Making_Payment_Country Total_Amount_of_Payment_USDollars Date_of_Payment Number_of_Payments_Included_in_Total_Amount Form_of_Payment_or_Transfer_of_Value Nature_of_Payment_or_Transfer_of_Value City_of_Travel State_of_Travel Country_of_Travel Physician_Ownership_Indicator Third_Party_Payment_Recipient_Indicator Name_of_Third_Party_Entity_Receiving_Payment_or_Transfer_of_Value Charity_Indicator Third_Party_Equals_Covered_Recipient_Indicator Contextual_Information Delay_in_Publication_Indicator Record_ID Dispute_Status_for_Publication Related_Product_Indicator Covered_or_Noncovered_Indicator_1 Indicate_Drug_or_Biological_or_Device_or_Medical_Supply_1 Product_Category_or_Therapeutic_Area_1 Name_of_Drug_or_Biological_or_Device_or_Medical_Supply_1 Associated_Drug_or_Biological_NDC_1 Associated_Device_or_Medical_Supply_PDI_1 Covered_or_Noncovered_Indicator_2 Indicate_Drug_or_Biological_or_Device_or_Medical_Supply_2 Product_Category_or_Therapeutic_Area_2 Name_of_Drug_or_Biological_or_Device_or_Medical_Supply_2 Associated_Drug_or_Biological_NDC_2 Associated_Device_or_Medical_Supply_PDI_2 Covered_or_Noncovered_Indicator_3 Indicate_Drug_or_Biological_or_Device_or_Medical_Supply_3 Product_Category_or_Therapeutic_Area_3 Name_of_Drug_or_Biological_or_Device_or_Medical_Supply_3 Associated_Drug_or_Biological_NDC_3 Associated_Device_or_Medical_Supply_PDI_3 Covered_or_Noncovered_Indicator_4 Indicate_Drug_or_Biological_or_Device_or_Medical_Supply_4 Product_Category_or_Therapeutic_Area_4 Name_of_Drug_or_Biological_or_Device_or_Medical_Supply_4 Associated_Drug_or_Biological_NDC_4 Associated_Device_or_Medical_Supply_PDI_4 Covered_or_Noncovered_Indicator_5 Indicate_Drug_or_Biological_or_Device_or_Medical_Supply_5 Product_Category_or_Therapeutic_Area_5 Name_of_Drug_or_Biological_or_Device_or_Medical_Supply_5 Associated_Drug_or_Biological_NDC_5 Associated_Device_or_Medical_Supply_PDI_5 Program_Year Payment_Publication_Date CHANGED Covered Recipient Teaching Hospital 220162 9984 Adventhealth Orlando null null null null null null 601 E Rollins St null Orlando FL 32803 United States null null null null null null null null null null null null null null null null null null null Integra LifeSciences Corporation 100000010950 Integra LifeSciences Corporation NJ United States 2000.00 07/20/2023 1 Cash or cash equivalent Grant null null null No No Third Party Payment null null null null No 1031636873 No Yes Covered Device Neuro Critical Care CODMAN CERTAS null 10381780529033 null null null null null null null null null null null null null null null null null null null null null null null null 2023 01/30/2025 CHANGED Covered Recipient Teaching Hospital 220162 9

CMS Covered Recipient File (https://download.cms.gov/openpayments/PHPRFL_P01302025_01212025.zip)

In [0]:
# File location and type
file_location = "/FileStore/tables/OP_CVRD_RCPNT_PRFL_SPLMTL_P01302025_01212025.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
OP_CVRD_df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(OP_CVRD_df.limit(5))

Covered_Recipient_Profile_Type Covered_Recipient_Profile_ID Associated_Covered_Recipient_Profile_ID_1 Associated_Covered_Recipient_Profile_ID_2 Covered_Recipient_NPI Covered_Recipient_Profile_First_Name Covered_Recipient_Profile_Middle_Name Covered_Recipient_Profile_Last_Name Covered_Recipient_Profile_Suffix Covered_Recipient_Profile_Alternate_First_Name Covered_Recipient_Profile_Alternate_Middle_Name Covered_Recipient_Profile_Alternate_Last_Name Covered_Recipient_Profile_Alternate_Suffix Covered_Recipient_Profile_Address_Line_1 Covered_Recipient_Profile_Address_Line_2 Covered_Recipient_Profile_City Covered_Recipient_Profile_State Covered_Recipient_Profile_Zipcode Covered_Recipient_Profile_Country_Name Covered_Recipient_Profile_Province_Name Covered_Recipient_Profile_Primary_Specialty Covered_Recipient_Profile_OPS_Taxonomy_1 Covered_Recipient_Profile_OPS_Taxonomy_2 Covered_Recipient_Profile_OPS_Taxonomy_3 Covered_Recipient_Profile_OPS_Taxonomy_4 Covered_Recipient_Profile_OPS_Taxonomy_5 Covered_Recipient_Profile_OPS_Taxonomy_6 Covered_Recipient_Profile_License_State_Code_1 Covered_Recipient_Profile_License_State_Code_2 Covered_Recipient_Profile_License_State_Code_3 Covered_Recipient_Profile_License_State_Code_4 Covered_Recipient_Profile_License_State_Code_5 Covered Recipient Physician 1 null null 1003020595 BARRY D SIMMONS null BARRON DAVID SIMMONS null 847 SOUTH MILLEDGE AVE null ATHENS GA 30605-1331 UNITED STATES null Dental Providers|Dentist|General Practice 1223G0001X null null null null null GA null null null null Covered Recipient Physician 2 null null 1003022823 LINDSEY R SCHUSTER null LINDSEY R GROBER null 15 HEIDI LN null MOUNT SINAI NY 11766-1428 UNITED STATES null Allopathic & Osteopathic Physicians|Emergency Medicine 207P00000X null null null null null NY PA null null null Covered Recipient Physician 3 null null 1003023359 JUSTIN H RACKLEY null JUSTIN HAMWAY RACKLEY null 1001 SAM PERRY BLVD null FREDERICKSBURG VA 22401-4453 UNITED STATES null Allopathic & Osteopathic Physicians|Anesthesiology 207L00000X null null null null null VA OH null null null Covered Recipient Physician 4 null null 1003024811 LISA L HAMAKER null null null null null 100 MARKET ST STE 300 null COLLEGEVILLE PA 19426-4927 UNITED STATES null Allopathic & Osteopathic Physicians|Internal Medicine|Endocrinology, Diabetes & Metabolism 207RE0101X null null null null null PA null null null null Covered Recipient Physician 5 null null 1003025875 PATRICK E DAVOL null PATRICK ERIN DAVOL null 1698 E MCANDREWS RD STE 280 null MEDFORD OR 97504-5590 UNITED STATES null Allopathic & Osteopathic Physicians|Urology 208800000X null null null null null OR IN null null null

**1. What is the Nature of Payments with reimbursement amounts greater than $1,000 ordered by count?**

In [0]:
df = df.withColumn("Total_Amount_of_Payment_USDollars", col("Total_Amount_of_Payment_USDollars").cast(IntegerType()))
high_reimbursement_df = df.filter(col("Total_Amount_of_Payment_USDollars") > 1000)
payments_df = high_reimbursement_df.groupBy("Nature_of_Payment_or_Transfer_of_Value") \
    .count() \
    .orderBy(desc("count"))
    
display(payments_df)

Nature_of_Payment_or_Transfer_of_Value count Compensation for services other than consulting, including serving as faculty or as a speaker at a venue other than a continuing education program 164092 Consulting Fee 105228 Travel and Lodging 24738 Honoraria 13750 Education 13376 Royalty or License 11534 Compensation for serving as faculty or as a speaker for a medical education program 8658 Grant 4922 Space rental or facility fees (teaching hospital only) 4917 Long term medical supply or device loan 2930 Debt forgiveness 1787 Food and Beverage 966 Gift 630 Acquisitions 562 Charitable Contribution 239 Entertainment 30

**Interpretation:** This analysis examines high-value reimbursement payments made to healthcare providers. It first casts the Total_Amount_of_Payment_USDollars column to an integer type and filters for payments exceeding $1,000. The dataset is then grouped by Nature_of_Payment_or_Transfer_of_Value, counting occurrences of each payment type and ordering the results in descending order of count. The output reveals that the most frequent high-value payment category is "Compensation for services other than consulting, including serving as faculty or as a speaker" (164,092 instances), followed by "Consulting Fees" (105,228 instances) and "Travel and Lodging" (24,738 instances). Other notable categories include "Honoraria," "Education," and "Royalty or License" payments. Less common payment types, such as "Food and Beverage" and "Gift," appear less frequently, likely because they involve smaller transactions that do not meet the $1,000 threshold.

**2. What are the top ten Nature of Payments by count?**

In [0]:
data = df.dropna(subset=["Nature_of_Payment_or_Transfer_of_Value"])

top_payments_df = data.groupBy("Nature_of_Payment_or_Transfer_of_Value") \
    .count() \
    .orderBy(desc("count"))\
    .limit(10)

display(top_payments_df)

Nature_of_Payment_or_Transfer_of_Value count Food and Beverage 13378464 Travel and Lodging 545086 Compensation for services other than consulting, including serving as faculty or as a speaker at a venue other than a continuing education program 236628 Consulting Fee 170630 Education 161078 Gift 31786 Honoraria 20232 Royalty or License 15865 Compensation for serving as faculty or as a speaker for a medical education program 12234 Entertainment 7967

**Interpretation:** This analysis examines the most common types of payments made to healthcare providers by grouping and counting the occurrences of each "Nature_of_Payment_or_Transfer_of_Value" category. The dataset is then grouped by payment type, and the number of instances for each type is counted. The results are ordered in descending order based on count. The output shows that "Food and Beverage" is the most frequent category, with over 13.3 million transactions, followed by "Travel and Lodging" (545,086 transactions). This suggests that companies frequently cover meal and travel expenses for healthcare providers. Other notable payment categories include "Compensation for services other than consulting" (236,628 instances) and "Consulting Fees" (170,630 instances), highlighting direct financial interactions. "Education," "Gifts," and "Honoraria" are also significant but less frequent. The presence of "Entertainment" (7,967 instances) suggests that some payments involve leisure-related activities.

**3. What are the top ten Nature of Payments by total amount?**

In [0]:
data_df = df.dropna(subset=["Nature_of_Payment_or_Transfer_of_Value"])

payment_amount_df = data_df.groupBy("Nature_of_Payment_or_Transfer_of_Value") \
    .agg(sum("Total_Amount_of_Payment_USDollars").alias("Total amount($)")) \
    .orderBy(desc("Total amount($)"))\
    .limit(10)

display(payment_amount_df)

Nature_of_Payment_or_Transfer_of_Value Total amount($) Royalty or License 1192167666 Compensation for services other than consulting, including serving as faculty or as a speaker at a venue other than a continuing education program 594620402 Consulting Fee 514840970 Food and Beverage 367908867 Travel and Lodging 179315496 Grant 111887837 Acquisitions 71925477 Education 64629090 Honoraria 55850335 Long term medical supply or device loan 30095899

**Interpretation:** This analysis examines the total dollar amount of payments made to healthcare providers across different "Nature_of_Payment_or_Transfer_of_Value" categories. The dataset is then grouped by payment type, and the total amount paid for each category is calculated using the .agg(sum(...)) function. The results are sorted in descending order by total payment amount. The output reveals that "Royalty or License" payments account for the largest total amount, exceeding 1.19 billion dollars, indicating significant financial transactions related to intellectual property. "Compensation for services other than consulting" (594.6 million dollars) and "Consulting Fees" (514.8 million dollars) also represent substantial payments, highlighting direct financial incentives for physicians. Interestingly, "Food and Beverage" transactions, which were the most frequent in the previous analysis, rank fourth in total amount (367.9 million dollars), suggesting that while these transactions are common, they typically involve smaller dollar values. Other notable payment categories include "Travel and Lodging" (179.3 million dollars), "Grants" ($111.9 million), and "Acquisitions" ($71.9 million).

**4. What are the top ten physician specialties by total amount?**

In [0]:
from pyspark.sql import functions as F
from pyspark.sql.types import LongType

merged_df = df.na.drop(subset=["Covered_Recipient_Profile_ID", "Total_Amount_of_Payment_USDollars"]) \
    .join(OP_CVRD_df, on="Covered_Recipient_Profile_ID", how="inner")
specialties_amount_df = merged_df.groupBy("Covered_Recipient_Specialty_1") \
    .agg(F.sum("Total_Amount_of_Payment_USDollars").cast(LongType()).alias("Total amount($)")) \
    .orderBy(F.desc("Total amount($)")) \
    .limit(10)

display(specialties_amount_df)


Covered_Recipient_Specialty_1 Total amount($) Allopathic & Osteopathic Physicians|Orthopaedic Surgery 403450213 Allopathic & Osteopathic Physicians|Internal Medicine 131363003 Allopathic & Osteopathic Physicians|Psychiatry & Neurology|Neurology 89792136 Allopathic & Osteopathic Physicians|Neurological Surgery 86088478 Allopathic & Osteopathic Physicians|Dermatology 83202647 Allopathic & Osteopathic Physicians|Internal Medicine|Cardiovascular Disease 70220841 Allopathic & Osteopathic Physicians|Internal Medicine|Hematology & Oncology 69468941 Allopathic & Osteopathic Physicians|Orthopaedic Surgery|Adult Reconstructive Orthopaedic Surgery 66772837 Allopathic & Osteopathic Physicians|Psychiatry & Neurology|Psychiatry 63298254 Allopathic & Osteopathic Physicians|Surgery 62493627

**Interpretation:** This analysis examines the top 10 physician specialties by total payment amount by first merging the physician profile data (OP_CVRD_df) with the payment transaction data (df) using the Covered_Recipient_Profile_ID. It ensures that any rows with missing values in critical columns (Covered_Recipient_Profile_ID and Total_Amount_of_Payment_USDollars) are excluded. The dataset is then grouped by physician specialty (Covered_Recipient_Specialty_1), and the total payment for each specialty is aggregated using the sum() function. The results are sorted in descending order, and the top 10 specialties are selected. The output shows that "Orthopaedic Surgery" receives the highest total payment, followed by specialties like "Internal Medicine," "Neurology," and "Dermatology." 

**5. Who are the top ten physicians by total amount?**


In [0]:
from pyspark.sql.types import LongType

OP_CVRD_df = OP_CVRD_df.selectExpr(
    "Covered_Recipient_Profile_ID", 
    "Covered_Recipient_Profile_First_Name", 
    "Covered_Recipient_Profile_Middle_Name", 
    "Covered_Recipient_Profile_Last_Name", 
    "Covered_Recipient_NPI", 
    "Covered_Recipient_Profile_Address_Line_1",
    "Covered_Recipient_Profile_City", 
    "Covered_Recipient_Profile_State", 
    "Covered_Recipient_Profile_Zipcode"
).distinct()

physicians_amount_df = df.na.drop(subset=["Covered_Recipient_Profile_ID", "Total_Amount_of_Payment_USDollars"]) \
    .groupBy("Covered_Recipient_Profile_ID") \
    .agg(sum("Total_Amount_of_Payment_USDollars").cast(LongType()).alias("Total amount($)")) \
    .limit(10)
physicians_amount_df = physicians_amount_df.join(OP_CVRD_df, on="Covered_Recipient_Profile_ID", how="inner") \
    .orderBy(desc("Total amount($)"))

display(physicians_amount_df)

Covered_Recipient_Profile_ID Total amount($) Covered_Recipient_Profile_First_Name Covered_Recipient_Profile_Middle_Name Covered_Recipient_Profile_Last_Name Covered_Recipient_NPI Covered_Recipient_Profile_Address_Line_1 Covered_Recipient_Profile_City Covered_Recipient_Profile_State Covered_Recipient_Profile_Zipcode 1396828 3109 MICHAEL BENJAMIN KRANTZOW 1306167614 9970 CENTRAL PARK BLVD N BOCA RATON FL 33428-2231 83101 1891 ULRICH null LUFT 1831115229 100 SPALDING DR NAPERVILLE IL 60540-6550 186264 1555 RICARDO null IZQUIERDO 1093821753 2425 W 22ND ST OAK BROOK IL 60523-1245 728183 740 RITA null HUTCHINSON BLACK 1801225255 7300 SANDLAKE COMMONS BLVD STE 220 ORLANDO FL 32819-8011 783782 389 CHRISTOPHER S JONES 1760514467 320 N MERIDIAN ST INDIANAPOLIS IN 46204-1719 10500007 370 THARA null BHANGU 1588091243 7600 SW 87TH AVE STE 206 MIAMI FL 33173 47919 298 ROBIN E HAUCK 1043445745 245 E MAIN ST RAMSEY NJ 07446-1942 10641949 271 REBECCA null CORBIN 1467040329 2637 BROADWAY NEW YORK NY 10025 476559 42 JEFFREY null NYMAN 1962419036 1200 W GODFREY AVE PHILADELPHIA PA 19141-3323 11461381 13 CHRISTOPHER null LE 1700560703 301 UNIVERSITY BLVD GALVESTON GALVESTON TX 77555-1385

**Interpretation:** I identified the top 10 physicians who received the highest total payments and retrieved their profile details. First, I selected relevant columns from OP_CVRD_df, ensuring distinct records of physicians’ profiles, including their names, NPI, and address details. Then, I processed df, which contains payment transaction data, by filtering out missing values in the Covered_Recipient_Profile_ID and Total_Amount_of_Payment_USDollars columns. The data was then grouped by Covered_Recipient_Profile_ID. The top 10 physicians with the highest total payments were selected and joined with their profile details using an inner join. The results were sorted in descending order based on payment amounts and displayed. The output shows that Michael Benjamin Krantzow received the highest total payment, followed by Ulrich Luft and Ricardo Izquierdo, with payments ranging from 3,109dollars to as low as 13dollars. 